src/nodes/header.rs

```diff
@@ -0,0 +1,407 @@
+        unsafe {
+            // SAFETY: This function is called when mismatch happened and
+            // we used the node to match the number of bytes,
+            // by this we know that len < prefix len, but since we + 1,
+            // to skip the key byte we have that len <= prefix len
+            assume!(end <= self.prefix.len());
+
+            // SAFETY: This is by construction end = begin + len
+            assume!(begin <= end);
+        }
+        self.prefix.copy_within(begin..end, 0);
+    }
+
+    /// Set the length of the prefix to 0 and returns a copy of the
+    /// prefix, length and capped length
+    #[inline(always)]
+    fn clear_prefix(&mut self) -> ([u8; NUM_PREFIX_BYTES], usize, usize) {
+        let len = self.prefix_len();
+        let capped_len = self.capped_prefix_len();
+        self.prefix_len = 0;
+
+        (self.prefix, len, capped_len)
+    }
+
+    /// Append `new` to the prefix and sums `new_len` to the prefix length
+    #[inline(always)]
+    fn push_prefix(&mut self, new: &[u8], new_len: usize) {
+        let begin = self.capped_prefix_len();
+        let end = (begin + new.len()).min(NUM_PREFIX_BYTES);
+        let len = end - begin;
+        self.prefix[begin..end].copy_from_slice(&new[..len]);
+        self.prefix_len += new_len as u32;
+    }
+
+    /// Increments the number of children
+    #[inline(always)]
+    fn inc_num_children(&mut self) {
+        self.num_children += 1;
+    }
+
+    /// Decrements the number of children
+    #[inline(always)]
+    fn dec_num_children(&mut self) {
+        self.num_children -= 1;
+    }
+}
+
+macro_rules! define_common_node_header_methods {
+    () => {
+        #[inline(always)]
+        fn read_prefix(&self) -> &[u8] {
+            &self.0.read_prefix()
+        }
+
+        #[inline(always)]
+        fn prefix_len(&self) -> usize {
+            self.0.prefix_len()
+        }
+
+        #[inline(always)]
+        fn capped_prefix_len(&self) -> usize {
+            self.0.capped_prefix_len()
+        }
+
+        #[inline(always)]
+        fn num_children(&self) -> usize {
+            self.0.num_children()
+        }
+
+        #[inline(always)]
+        fn ltrim_by(&mut self, len: usize) {
+            self.0.ltrim_by(len)
+        }
+
+        #[inline(always)]
+        fn clear_prefix(&mut self) -> ([u8; NUM_PREFIX_BYTES], usize, usize) {
+            self.0.clear_prefix()
+        }
+
+        #[inline(always)]
+        fn inc_num_children(&mut self) {
+            self.0.inc_num_children()
+        }
+
+        #[inline(always)]
+        fn dec_num_children(&mut self) {
+            self.0.dec_num_children()
+        }
+
+        #[inline(always)]
+        fn push_prefix(&mut self, new: &[u8], new_len: usize) {
+            self.0.push_prefix(new, new_len)
+        }
+    };
+}
+
+/// Specifies the behaviour of different headers types
+pub trait NodeHeader<const NUM_PREFIX_BYTES: usize>: Debug + Clone + PartialEq + Eq {
```

Comment:
I'm a little doubtful that this extra complexity of:
1. generic header type
2. generic NUM_PREFIX_BYTES

is entirely worth it. Did you happen to capture performance changes for adding this bit? I also know of an existing crate with implements a fix-key length ART: https://docs.rs/art/latest/art/.

Do you think it would better just to defer to that crate for fix-length key situations rather than add to the complexity of this implementation?

Reply:
Yeah I also didn't like the complexity of the generics, but when I tested memory consumption and performance with integers it made a big difference. So IDK, what do you think ? Should we remove it ?

Reply:
Can you provide some of the numbers for the performance difference?